extensions/amp-ad/0.1/amp-ad-api-handler.js

```diff
@@ -121,21 +129,31 @@ export class AmpAdApiHandler {
       // creative as it will not expect having to send the render-start message.
       this.iframe_.style.visibility = 'hidden';
     }
-    listenForOnce(this.iframe_, 'render-start', () => {
-      if (!this.iframe_) {
-        return;
-      }
-      if (this.baseInstance_.renderStartResolve_) {
-        this.baseInstance_.renderStartResolve_();
-        this.baseInstance_.renderStartResolve_ = null;
-      }
-      this.iframe_.style.visibility = '';
-    }, this.is3p_);
+
+    this.renderStartPromise_ =
+        listenForOncePromise(this.iframe_, 'render-start', this.is3p_);
+
     this.viewer_.onVisibilityChanged(() => {
       this.sendEmbedInfo_(this.baseInstance_.isInViewport());
     });
     this.element_.appendChild(this.iframe_);
-    return loadPromise(this.iframe_);
+    return loadPromise(this.iframe_).then(() => {
+      return timerFor(this.baseInstance_.win)
+          .timeoutPromise(200, this.renderStartPromise_,
+          'render-start-event timed out')
+          .catch(error => {
+            //TODO: report performance
+            if (this.iframe_) {
+              this.iframe_.style.visibility = '';
+            }
+            throw error;
```

Comment:
never receive render-start, I think we should fail the startUp promise and throw the error to `layoutCallback`